build_rules.gradle

```diff
@@ -0,0 +1,275 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * License); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an AS IS BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+println "Applying build_rules.gradle to $project.name"
+
+/*************************************************************************************************/
+// Apply common properties/repositories and tasks to all builds.
+
+group = "org.apache.beam"
+version = "2.3.0-SNAPSHOT"
+
+// Define the default set of properties
+repositories {
+  mavenLocal()
+  mavenCentral()
+  jcenter()
+
+  // Release staging repository
+  maven { url "https://oss.sonatype.org/content/repositories/staging/" }
+
+  // Apache nightly snapshots
+  maven { url "https://repository.apache.org/snapshots" }
+
+  // Apache release snapshots
+  maven { url "https://repository.apache.org/content/repositories/releases" }
+}
+
+// Provide code coverage
+apply plugin: "jacoco"
+
+// Provies "htmlDependencyReport" task which spits out a nice page per subproject
+// showing all the dependencies per Gradle configuration.
+apply plugin: "project-report"
+
+// Apply a task dependency visualization plugin which creates a ".dot" file in the build directory
+// giving the task dependencies for that project
+apply plugin: "cz.malohlava.visteg"
+
+/*************************************************************************************************/
+
+/** Used as configuration within the applyJavaNature closure. */
+class JavaNatureConfiguration {
```

Comment:
<!--new_thread; commit:8ca3fdca276a74973cd428d7321b995f7b7ebccf; resolved:0-->
To be clear, this is our own POJO and the fields are just for us to go ahead and use? Or are these magic fields that some plugin looks at?

Reply:
Our own configuration, this allows us to call applyJavaNature with named arguments instead of positional arguments. This is done by having applyJavaNature accept a map and convert that map to an object.